samza-sql-avro/src/main/java/org/apache/samza/sql/avro/AvroRelToSamzaMsgConverterFactory.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.avro.reflect.ReflectData;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverter;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverterFactory;
+
+
+/**
+ * Factory that creates an object of {@link AvroRelToSamzaMsgConverter}, which is used to convert the Relational messages
+ * into avro messages using the avro schema passed. If the avro schema is not passed for the conversion, This converter
+ * will generate the schema based on the fields and types of the fields present in the {@link SamzaSqlRelMessage}.
+ */
+public class AvroRelToSamzaMsgConverterFactory implements RelToSamzaMsgConverterFactory {
+
+  private Map<String, Schema> schemas = new HashMap<>();
+
+  @Override
+  public RelToSamzaMsgConverter create(Properties config) {
+    return new AvroRelToSamzaMsgConverter(config);
+  }
+
+
+  private class AvroRelToSamzaMsgConverter implements RelToSamzaMsgConverter {
+
+    private final Schema arraySchema;
+    private final Schema mapSchema;
+    private Map<String, Schema> streamToSchemaCache = new HashMap<>();
+
+    public class TestClass {
+      public List<Object> arrayValue = new ArrayList<>();
+      public Map<String, Object> mapValue = new HashMap<>();
+    }
+
+    public AvroRelToSamzaMsgConverter(Properties config) {
+      Schema testClassSchema = ReflectData.get().getSchema(TestClass.class);
+      arraySchema = testClassSchema.getField("arrayValue").schema();
+      mapSchema = testClassSchema.getField("mapValue").schema();
+    }
+
+    @Override
+    public Object extactKey(String streamName, Optional<String> schema, SamzaSqlRelMessage relMessage) {
+      return relMessage.getField(AvroSamzaToRelMsgConverterFactory.KEY_NAME).orElse(null);
+    }
+
+    @Override
+    public Object convert(String streamName, Optional<String> schemaValue, SamzaSqlRelMessage relMessage) {
+      String schemaStr = schemaValue.orElseGet(() -> computeSchema(streamName, relMessage));
+      Schema schema;
+      // See whether we have a schema object corresponding to the schemaValue and reuse it.
+      // CachedSchemaRegistryClient doesn't like if we recreate schema objects.
+      if (schemas.containsKey(schemaStr)) {
+        schema = schemas.get(schemaStr);
+      } else {
+        schema = Schema.parse(schemaStr);
+        schemas.put(schemaStr, schema);
+      }
+
+      GenericRecord record = new GenericData.Record(schema);
+      String[] fieldNames = relMessage.getNames();
+      Object[] values = relMessage.getValue();
+      for (int index = 0; index < fieldNames.length; index++) {
+        if (fieldNames[index].equals(AvroSamzaToRelMsgConverterFactory.KEY_NAME) || values[index] == null) {
+          continue;
+        }
+        record.put(fieldNames[index], values[index]);
+      }
+
+      return record;
+    }
+
+    private String computeSchema(String streamName, SamzaSqlRelMessage relMessage) {
```

Comment:
@nickpan47 What you are describing above is DDL. When you do a 
`select foo,bar from testMysqlTable` 
The output of this query is list of rows, where each row is list of columns. This is what is being represented in a SamzaSqlRelMessage. Each SamzaSqlRelMessage represents the row in the relational output. SamzaSqlRelMessage doesn't  support any additional metadata about the columns *yet*, mainly because I haven't found the need for it.

If your question is how do convert this relational message into a Samza message which is <key, value> pair where value is a serialized record of any arbitrary format ? This is controlled by pluggable RelToSamzaConverter, In the AvroRelToSamzaConverter it uses the schema to create the value and it uses the special fieldName "__system_key" to fill up the key part of the samza message.